inc/templates.php

```diff
@@ -274,3 +279,159 @@ function convert_blocks_to_components( array $blocks ): array {
 
 	return $components;
 }
+
+/**
+ * Recursively iterate on a component tree.
+ *
+ * @param array $components Array of components.
+ * @return array
+ */
+function traverse_components( array $components ): array {
```

Comment:
I think we can iterate on the hydration logic here, but this looks like a good start. I think it would be helpful for us to build some basic unit tests out of a few basic use cases to ensure that this is doing what we expect and that we don't introduce regressions as we mature this functionality.

Generally speaking, it would be nice if we could get this down to a process wherein hydration is done in a few contained steps:

1. Resolve any template partials to construct a template tree
2. Recursively loop through template tree
3. Instantiate `WP_Component` objects for each component found, passing configuration from the template to the constructor of the object in order to set up the right data.
4. Return hydrated data objects using the callback function from each registered component in the tree.

Reply:
I agree, and I'd like to merge what we have for now and revisit to improve the base logic.